lib/parsers/webhook_json_parser.rb

```diff
@@ -5,35 +5,96 @@ module Sinatra
   module Parsers
     class WebhookJsonParser
       def call(body)
-        data = JSON.parse(body, quirks_mode: true)
+        @data = JSON.parse(body, quirks_mode: true)
+        @vcs  = detect_vcs
         {
-          branch: parse_branch(data),
-          deleted: parse_deleted?(data)
+          branch:  branch,
+          deleted: deleted?
         }
       end
 
-      def parse_branch(data)
-        (
-          data['ref'] rescue nil || # github & gitlab
-          data['refChanges'][0]['refId'] rescue nil ||  # stash
-          data['push']['changes'][0]['new']['name'] rescue nil ||  # bitbucket
-          data['resource']['refUpdates'][0]['name'] rescue nil ||  # TFS/VisualStudio-Git
-          data['repository']['default_branch'] rescue nil # github tagged release; no ref.
-        ).sub('refs/heads/', '') rescue nil
-      end
-
-      def parse_deleted?(data)
-        branch_deleted = (
-          data['deleted'] rescue nil || # Github
-          data['push']['changes'][0]['closed'] rescue nil || # Bitbucket
-          data['refChanges'][0]['type'] rescue nil || # Stash/Bitbucket Server
-          data['after'] rescue nil # Gitlab
-        )
-        [
-          true,
-          'DELETED',
-          '0000000000000000000000000000000000000000'
-        ].include?(branch_deleted)
+      def detect_vcs
+        return 'github'    if github_webhook?
+        return 'gitlab'    if gitlab_webhook?
+        return 'stash'     if stash_webhook?
+        return 'bitbucket' if bitbucket_webhook?
+        return 'tfs'       if tfs_webhook?
+        raise StandardError, 'payload not recognised'
+      end
+
+      def github_webhook?
+        # https://developer.github.com/v3/activity/events/types/#pushevent
+        # X-GitHub-Event header is set, but not accessible here.
+        return false unless @data.key? 'repository'
```

Comment:
I feel like the logic should be reversed: ```return false``` should be the default Maybe easier:

```
@data['repository']['html_url'] =~ /github\.com/ rescue false
```

However, this would break github enterprise which runs on a different URL.

Reply:
@ekohl If we can avoid continuing to use the `data contains expressions rescue blah` format, I'd prefer that. It's against the style guide anyway and (like Bitbucket and Bitbucket Server) we may just have to create a separate method for checking github enterprise.

Thoughts?

Reply:
I'd suggest different parsers for every service and do header detection beforehand if possible

Reply:
Hmm, that might work. I'm still a bit new to Rack::Parser, but they would probably have to be other classes or modules called by `Sinatra::Parsers::WebhookJsonParser`. I'm not sure `Rack::Parser` allows you to define multiple custom parsers for a single Content-Type

Reply:
I would have liked to use the headers, but they're not passed to us.
https://github.com/achiu/rack-parser/blob/master/lib/rack/parser.rb#L28

Reply:
I wonder if there is an easy way to parse the Headers and set the parser based on that.

Reply:
I say we merge this PR and then move to iterate on what @alexjfisher has done by having `Sinatra::Parsers::WebhookParser` parse the headers to determine source service then pass the data to sub-parsers for those services.

Trying to parse headers up front or reformat the `Content-Type` is overly complex for what we really need and may cause regret in the future.

Reply:
I'd be fine with that.

Reply:
After additional research, I've found that if we use `Rack::Parser`, we do not have access to the request headers _before_ `Rack::Parser` sets a custom parser and headers are not available within the custom parser either as `Rack::Parser` discards/doesn't make them available to use to use.

We will either have to:

1. Build our own middleware for parsing both headers and body content.
2. Not use headers at all for determining content source.
3. Get rid of middleware altogether and use our own classes that parse the data into what we need and pass it back to the routes.

thoughts? Anything I missed?